examples/ecs/fallible_system_param.rs

```diff
@@ -0,0 +1,245 @@
+        .insert_resource(CurrentGameMode::new(Deathmatch {
+            max_time: 600.0,
+            max_kills: 30,
+        }))
+        .add_systems(Update, update_deathmatch)
+        .run();
+}
+
+// This struct encapsulates the trait object so it can used as a resource.
+#[derive(Resource)]
+pub struct CurrentGameMode(Box<dyn GameMode>);
+
+impl CurrentGameMode {
+    pub fn new(mode: impl GameMode) -> Self {
+        Self(Box::new(mode))
+    }
+
+    pub fn to_ref<T: GameMode>(&self) -> Option<&T> {
+        GameMode::as_reflect(&*self.0).downcast_ref()
+    }
+
+    pub fn to_mut<T: GameMode>(&mut self) -> Option<&mut T> {
+        GameMode::as_reflect_mut(&mut *self.0).downcast_mut()
+    }
+
+    pub fn type_name(&self) -> String {
+        self.0.type_name().to_string()
+    }
+}
+
+// This is the error type returned when the resource isn't the correct type.
+#[derive(Debug)]
+pub enum GameError {
+    MissingResource,
+    WrongType(&'static str, String),
+}
+
+impl std::fmt::Display for GameError {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            Self::MissingResource => write!(f, "Resource `CurrentGameMode` missing"),
+            Self::WrongType(expected, found) => {
+                write!(f, "Expected game mode `{expected}`, found `{found}`")
+            }
+        }
+    }
+}
+
+impl std::error::Error for GameError {}
+
+impl From<ResError<CurrentGameMode>> for GameError {
+    fn from(_: ResError<CurrentGameMode>) -> Self {
+        Self::MissingResource
+    }
+}
+
+impl From<ResMutError<CurrentGameMode>> for GameError {
+    fn from(_: ResMutError<CurrentGameMode>) -> Self {
+        Self::MissingResource
+    }
+}
+
+// This is the optional system param that abstracts away converting the resource to the actual type.
+pub struct Game<'w, T: GameMode> {
+    mode: &'w T,
+}
+
+// Deref makes it convenient to interact with the actual data.
+impl<'w, T: GameMode> Deref for Game<'w, T> {
+    type Target = T;
+
+    fn deref(&self) -> &Self::Target {
+        self.mode
+    }
+}
+
+#[doc(hidden)]
+pub struct GameState {
+    // The `OptionalSystemParam::State` of `Res<CurrentGameMode>`.
+    mode_id: ComponentId,
+}
+
+// SAFETY: The resource state is initialized.
+unsafe impl<'w, T: GameMode> FallibleSystemParam for Game<'w, T> {
+    type State = GameState;
+    type Item<'world, 'state> = Game<'world, T>;
+    type Error = GameError;
+
+    fn init_state(world: &mut World, system_meta: &mut SystemMeta) -> Self::State {
+        GameState {
+            mode_id: <Res<CurrentGameMode> as FallibleSystemParam>::init_state(world, system_meta),
+        }
+    }
+
+    unsafe fn get_param<'world>(
+        state: &mut Self::State,
+        system_meta: &SystemMeta,
+        world: &'world World,
+        change_tick: Tick,
+    ) -> Result<Game<'world, T>, GameError> {
+        let current_mode = <Res<CurrentGameMode> as FallibleSystemParam>::get_param(
```

Comment:
```suggestion
        // SAFETY: This type has the same safety invariants as `Res<CurrentGameMode>`,
        // so the caller of this function must ensure that the safety invariants of `Res::get_param`
        // are upheld.
        let current_mode = <Res<CurrentGameMode> as FallibleSystemParam>::get_param(
```

Since this is an example, I think it's good to document each unsafe call, even the trivial ones.